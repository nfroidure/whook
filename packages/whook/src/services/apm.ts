import { name, autoService, location } from 'knifecycle';
import { type LogService } from 'common-services';
import { type JsonValue } from 'type-fest';

export type WhookAPMDependencies = {
  log: LogService;
};
export type WhookAPMService<T extends string = string> = (
  type: T,
  data: JsonValue,
) => void;

export default location(name('apm', autoService(initAPM)), import.meta.url);

const noop = () => undefined;

/**
 * Application monitoring service that simply log stringified contents.
 * @function
 * @param  {Object}     services
 * The services to inject
 * @param  {Function}   [services.log]
 * A logging function
 * @return {Promise<Object>}
 * A promise of the apm service.
 */
async function initAPM<T extends string = string>({
  log = noop,
}: WhookAPMDependencies): Promise<WhookAPMService<T>> {
  log('debug', '❤️ - Initializing the APM service.');

  return (type, data) => log('info', type, JSON.stringify(data));
}
