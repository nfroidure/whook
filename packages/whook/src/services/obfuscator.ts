import { autoService, location } from 'knifecycle';
import { type LogService } from 'common-services';
import { type JsonValue } from 'type-fest';
import { type WhookHeaders } from '../index.js';

export type WhookSensibleValueDescriptor = {
  name: string;
  pattern: RegExp;
  clearIndices: number[];
};
export type WhookObfuscatorConfig = {
  SHIELD_CHAR?: string;
  MAX_CLEAR_CHARS?: number;
  MAX_CLEAR_RATIO?: number;
  SENSIBLE_PROPS?: WhookSensibleValueDescriptor[];
  SENSIBLE_HEADERS?: WhookSensibleValueDescriptor[];
};
export type WhookObfuscatorDependencies = {
  log?: LogService;
} & WhookObfuscatorConfig;
export type WhookObfuscatorService = {
  obfuscate: (secret: string) => string;
  obfuscateSensibleProps: (
    propValue: JsonValue,
    propName?: string,
  ) => JsonValue;
  obfuscateSensibleHeaders: (headers: WhookHeaders) => WhookHeaders;
};

export default location(autoService(initObfuscator), import.meta.url);

const noop = () => undefined;

const DEFAULT_MAX_CLEAR_CHARS = 6;
const DEFAULT_MAX_CLEAR_RATIO = 4;
const DEFAULT_SHIELD_CHAR = 'üõ°';
const DEFAULT_SENSIBLE_HEADERS: WhookSensibleValueDescriptor[] = [
  {
    name: 'authorization',
    pattern: /^(Bearer |Basic )(.*)$/i,
    clearIndices: [0],
  },
  { name: 'cookie', pattern: /^(.*)$/i, clearIndices: [] },
  {
    name: 'set-cookie',
    pattern: /^([^=]+=)([^;]*)(.*)$/i,
    clearIndices: [0, 2],
  },
];
const DEFAULT_SENSIBLE_PROPS: WhookSensibleValueDescriptor[] = [
  {
    name: 'access_token',
    pattern: /^(.*)$/i,
    clearIndices: [0],
  },
  {
    name: 'refresh_token',
    pattern: /^(.*)$/i,
    clearIndices: [0],
  },
  {
    name: 'password',
    pattern: /^(.*)$/i,
    clearIndices: [],
  },
  {
    name: 'token',
    pattern: /^(.*)$/i,
    clearIndices: [],
  },
  ...DEFAULT_SENSIBLE_HEADERS,
];

/**
 * Obfuscate sensible informations.
 * @param  {Object}   services
 * The service dependend on
 * @param  {Object}   [services.SHIELD_CHAR]
 * The char for replacing sensible informations
 * @param  {Object}   [services.MAX_CLEAR_CHARS]
 * The maximum clear chars to display
 * @param  {Object}   [services.MAX_CLEAR_RATIO]
 * The maximum clear chars ratio to display
 * @param  {Object}   [services.SENSIBLE_PROPS]
 * Sensible properties names
 * @param  {Object}   [services.SENSIBLE_HEADERS]
 * Sensible headers names
 * @return {Promise<Object>}
 * A promise of an object containing the gathered constants.
 * @example
 * import { initObfuscator } from '@whook/whook';
 * import { alsoInject } from 'knifecycle';
 * import { log } from 'node:console';
 *
 * const obfuscator = await initObfuscator();
 *
 * log(obfuscator('my very secret information!));
 * // my ...on!
 */
async function initObfuscator({
  SHIELD_CHAR = DEFAULT_SHIELD_CHAR,
  MAX_CLEAR_CHARS = DEFAULT_MAX_CLEAR_CHARS,
  MAX_CLEAR_RATIO = DEFAULT_MAX_CLEAR_RATIO,
  SENSIBLE_PROPS = DEFAULT_SENSIBLE_PROPS,
  SENSIBLE_HEADERS = DEFAULT_SENSIBLE_HEADERS,
  log = noop,
}: WhookObfuscatorDependencies): Promise<WhookObfuscatorService> {
  log('debug', 'üï∂Ô∏è - Initializing the obfuscator service.');

  return {
    obfuscate,
    obfuscateSensibleProps,
    obfuscateSensibleHeaders,
  };

  function obfuscate(secret) {
    const numClearChars = Math.min(
      MAX_CLEAR_CHARS,
      Math.floor(secret.length / MAX_CLEAR_RATIO),
    );

    if (numClearChars <= 1) {
      return SHIELD_CHAR;
    }

    return (
      secret.slice(0, Math.ceil(numClearChars / 2)) +
      '...' +
      secret.slice(secret.length - Math.floor(numClearChars / 2))
    );
  }

  function selectivelyObfuscateAll(
    pattern: RegExp,
    clearIndices: number[],
    values: string | string[],
  ) {
    return values instanceof Array
      ? values.map((value) =>
          selectivelyObfuscate(pattern, clearIndices, value),
        )
      : selectivelyObfuscate(pattern, clearIndices, values);
  }

  function selectivelyObfuscate(
    pattern: RegExp,
    clearIndices: number[],
    value: string,
  ) {
    // SECURITY: Here, we first test the pattern to ensure
    // the selective obfuscation will work, if not, we obfuscate
    // the whole value to default to security
    return 'string' !== typeof value
      ? obfuscate('')
      : pattern.test(value)
        ? value.replace(pattern, (...args) => {
            return args
              .slice(1, -2)
              .map((value, index) =>
                clearIndices.includes(index) ? value : obfuscate(value),
              )
              .join('');
          })
        : obfuscate(value);
  }

  function obfuscateSensibleHeaders(headers: WhookHeaders): WhookHeaders {
    return Object.keys(headers).reduce((finalHeaders, headerName) => {
      const sensibleHeader = SENSIBLE_HEADERS.find(
        (sensibleHeader) =>
          sensibleHeader.name.toLowerCase() === headerName.toLowerCase(),
      );
      return {
        ...finalHeaders,
        [headerName]: sensibleHeader
          ? selectivelyObfuscateAll(
              sensibleHeader.pattern,
              sensibleHeader.clearIndices,
              headers[headerName],
            )
          : headers[headerName],
      };
    }, {});
  }

  function obfuscateSensibleProps(
    propValue: JsonValue,
    propName = '_',
  ): JsonValue {
    if (propValue instanceof Array) {
      return propValue.map((value) => obfuscateSensibleProps(value, propName));
    } else if (typeof propValue === 'object' && propValue !== null) {
      return Object.keys(propValue).reduce(
        (newObject, key) => ({
          ...newObject,
          [key]: obfuscateSensibleProps(propValue[key] as JsonValue, key),
        }),
        {},
      );
    } else if (
      typeof propValue === 'boolean' ||
      typeof propValue === 'string' ||
      typeof propValue === 'number'
    ) {
      const sensibleProp = SENSIBLE_PROPS.find(
        (sensibleProp) =>
          sensibleProp.name.toLowerCase() === propName.toLowerCase(),
      );

      return sensibleProp
        ? selectivelyObfuscate(
            sensibleProp.pattern,
            sensibleProp.clearIndices,
            propValue.toString(),
          )
        : propValue;
    } else if (null === propValue) {
      return null;
    }
    return null;
  }
}
