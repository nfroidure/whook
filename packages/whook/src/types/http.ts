import { type JsonValue } from 'type-fest';
import { type Readable } from 'node:stream';

export type WhookHeaders = Record<string, string | string[]>;
export type WhookRequestBody = JsonValue | Readable;
export type WhookResponseBody = JsonValue | Readable;

export type WhookRequest = {
  url: string;
  method: string;
  headers: WhookHeaders;
  body?: WhookRequestBody;
};

export type WhookResponse = {
  status: number;
  headers?: WhookHeaders;
  body?: WhookResponseBody;
};
