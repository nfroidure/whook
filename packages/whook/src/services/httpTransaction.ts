import { service, location } from 'knifecycle';
import { YHTTPError } from 'yhttperror';
import statuses from 'statuses';
import ms from 'ms';
import { type WhookObfuscatorService } from './obfuscator.js';
import { type WhookAPMService } from './apm.js';
import { printStackTrace, YError } from 'yerror';
import {
  type LogService,
  type TimeService,
  type DelayService,
} from 'common-services';
import { type IncomingMessage, type ServerResponse } from 'node:http';
import { type JsonValue } from 'type-fest';
import { type Readable } from 'node:stream';
import { pickFirstHeaderValue } from '../libs/headers.js';
import { type WhookRequest, type WhookResponse } from '../types/http.js';

export type WhookHTTPTransactionConfig = {
  TIMEOUT?: number;
  TRANSACTIONS?: Record<string, Record<string, JsonValue>>;
};
export type HTTPTransactionDependencies = WhookHTTPTransactionConfig & {
  obfuscator: WhookObfuscatorService;
  delay: DelayService;
  log?: LogService;
  apm?: WhookAPMService;
  time?: TimeService;
  uniqueId?: () => string;
};

export type WhookHTTPTransaction = {
  request: WhookRequest;
  transaction: {
    id: string;
    start: (
      buildResponse: () => Promise<WhookResponse>,
    ) => Promise<WhookResponse>;
    catch: (err: Error) => Promise<void>;
    end: (response: WhookResponse, operationId?: string) => Promise<void>;
  };
};

export type WhookHTTPTransactionService = (
  req: IncomingMessage,
  res: ServerResponse,
) => Promise<WhookHTTPTransaction>;

const noop = () => undefined;
const DEFAULT_TIMEOUT = ms('30s');

function createIncrementor(n = 0) {
  return function increment() {
    return n++ + '';
  };
}

/* Architecture Note #2.10: HTTP Transactions

[Whook](https://github.com/nfroidure/whook) takes a very
 unusual direction when it comes to dealing with HTTP
 transactions.
It makes requests and responses serializable (thanks to
 `WhookRequest` and `WhookResponse` types) to:

- only work with functions that take request and return
 responses (allowing your handlers to be pure functions),
- have  easily unit testable handlers thanks to concise
 snapshots.

This service is intended to build those literal objects 
 from Node HTTP ones (famously known as req/res) before
 passing them to the handlers. It also keeps track of
 running queries and ensure it is well handled by the
 server before releasing it. If not, the transaction is
 resolved with an error response (for timeouts or when
 an error were caught).

The `httpTransaction` service creates a new transaction
 for every single HTTP request incoming. It helps
 ensuring each request receives a response and avoid
 idle requests via a configurable timeout.

It is also a convenient abstraction of the actual
 request/response between the router and
 the NodeJS world. A common need is to fake the
 HTTP method for backward compatibility with old
 browsers/proxies by using the
 `X-HTTP-Method-Override` header.

You can simply do this by wrapping this service. See
 the @whook/method-override` module for a working
 example.
 */
export default location(
  service(initHTTPTransaction, 'httpTransaction', [
    '?TIMEOUT',
    '?TRANSACTIONS',
    'delay',
    'obfuscator',
    '?log',
    '?apm',
    '?time',
    '?uniqueId',
  ]),
  import.meta.url,
);

/**
 * Instantiate the httpTransaction service
 * @function
 * @param  {Object}     services
 * The services to inject
 * @param  {Number}     [services.TIMEOUT=30000]
 * A number indicating how many ms the transaction
 *  should take to complete before being cancelled.
 * @param  {Object}     [services.TRANSACTIONS={}]
 * A hash of every current transactions
 * @param  {Object}     services.delay
 * A delaying service
 * @param  {Object}     services.obfuscator
 * A service to avoid logging sensible information
 * @param  {Function}   [services.log]
 * A logging function
 * @param  {Function}   [services.apm]
 * An apm function
 * @param  {Function}   [services.time]
 * A timing function
 * @param  {Function}   [services.uniqueId]
 * A function returning unique identifiers
 * @return {Promise<WhookHTTPTransaction>}
 * A promise of the httpTransaction function
 * @example
 * import initHTTPTransaction from '@whook/whook';
 * import { log } from 'node:console';
 *
 * const httpTransaction = await initHTTPTransaction({
 *   log,
 *   time: Date.now.bind(Date),
 * });
 */
async function initHTTPTransaction({
  TIMEOUT = DEFAULT_TIMEOUT,
  TRANSACTIONS,
  apm = noop,
  obfuscator,
  log = noop,
  time = Date.now.bind(Date),
  delay,
  uniqueId = createIncrementor(),
}: HTTPTransactionDependencies): Promise<WhookHTTPTransactionService> {
  // Not using default value to always
  // get an empty object here and avoid
  // conflicts between instances spawned
  // with defaults
  const FINAL_TRANSACTIONS: Record<
    string,
    Record<string, JsonValue>
  > = TRANSACTIONS || {};

  log('debug', '💱 - HTTP Transaction initialized.');

  return httpTransaction;

  /**
   * Create a new HTTP transaction
   * @function
   * @param  {HTTPRequest}  req
   * A raw NodeJS HTTP incoming message
   * @param  {HTTPResponse} res
   * A raw NodeJS HTTP response
   * @return {Array}
   * The normalized request and the HTTP
   * transaction created in an array.
   */
  async function httpTransaction(
    req: IncomingMessage,
    res: ServerResponse,
  ): Promise<WhookHTTPTransaction> {
    let initializationPromise;

    /* Architecture Note #2.10.1: New Transaction
    The idea is to maintain a hash of each pending
     transaction. To do so, we create a transaction
     object that contains useful information about
     the transaction and we store it into the
     `TRANSACTIONS` hash.

    Each transaction has a unique id that is either
     generated or picked up in the `Transaction-Id`
     request header. This allows to trace
     transactions end to end with that unique id.
    */
    const request = {
      url: req.url as string,
      method: (req.method as string).toLowerCase(),
      headers: Object.keys(req.headers).reduce(
        (finalHeaders, name) => ({
          ...finalHeaders,
          [name]: req.headers[name],
        }),
        {},
      ),
      body: req,
    };
    /**
    @typedef WhookHTTPTransaction
  */
    const transaction = {
      id: '',
      protocol: 'http',
      ip:
        '' +
          (pickFirstHeaderValue('x-forwarded-for', req.headers) || '').split(
            ',',
          )[0] ||
        req.socket.remoteAddress ||
        'unknown',
      startInBytes: req.socket.bytesRead,
      startOutBytes: req.socket.bytesWritten,
      startTime: time(),
      url: req.url as string,
      method: req.method as string,
      reqHeaders: obfuscator.obfuscateSensibleHeaders(request.headers),
      errored: false,
    };
    const delayPromise = delay.create(TIMEOUT);
    /**
     * Id of the transaction
     * @memberof WhookHTTPTransaction
     * @name id
     */
    let id: string =
      pickFirstHeaderValue('transaction-id', req.headers) || uniqueId();

    // Handle bad client transaction ids
    if (FINAL_TRANSACTIONS[id]) {
      initializationPromise = Promise.reject(
        new YHTTPError(400, 'E_TRANSACTION_ID_NOT_UNIQUE', id),
      );
      id = uniqueId();
    } else {
      initializationPromise = Promise.resolve();
    }

    transaction.id = id;
    FINAL_TRANSACTIONS[id] = transaction;

    return {
      request,
      transaction: {
        id,
        start: startTransaction.bind(
          null,
          { id, delayPromise },
          initializationPromise,
        ),
        catch: catchTransaction.bind(null, { id, req, res }),
        end: endTransaction.bind(null, { id, req, res, delayPromise }),
      },
    };
  }

  /**
   * Start the transaction
   * @memberof WhookHTTPTransaction
   * @name start
   * @param  {Function}   buildResponse
   * A function that builds a response
   * @return {Promise<Object>}
   * A promise to be resolved with the signed token.
   */
  async function startTransaction(
    { id, delayPromise }: { id: string; delayPromise: Promise<void> },
    initializationPromise: Promise<void>,
    buildResponse: () => Promise<WhookResponse>,
  ) {
    /* Architecture Note #2.10.2: Transaction start
  Once initiated, the transaction can be started. It
   basically spawns a promise that will be resolved
   to the actual response or rejected if the timeout
   is reached.
  */
    return Promise.race([
      initializationPromise.then(async () => buildResponse()),
      delayPromise.then(async () => {
        throw new YHTTPError(504, 'E_TRANSACTION_TIMEOUT', TIMEOUT, id);
      }),
    ]);
  }

  /**
   * Catch a transaction error
   * @memberof WhookHTTPTransaction
   * @name catch
   * @param  {Error}   err
   * A function that builds a response
   * @return {Promise}
   * A promise to be resolved with the signed token.
   */
  async function catchTransaction(
    { id, req }: { id: string; req: IncomingMessage },
    err: Error | YError | YHTTPError,
  ) {
    /* Architecture Note #2.10.3: Transaction errors
  Here we are simply logging errors.
   It is important for debugging but also for
   ending the transaction properly if an error
   occurs.
  */
    apm('ERROR', {
      guruMeditation: id,
      request:
        FINAL_TRANSACTIONS[id].protocol +
        '://' +
        (req.headers.host || 'localhost') +
        FINAL_TRANSACTIONS[id].url,
      verb: req.method as string,
      status: (err as YHTTPError).httpCode || 500,
      code: (err as YError).code || 'E_UNEXPECTED',
      stack: printStackTrace(err as Error),
      details: (err as YError).params || [],
    });

    FINAL_TRANSACTIONS[id].errored = true;

    throw err;
  }

  /**
   * End the transaction
   * @memberof WhookHTTPTransaction
   * @name end
   * @param  {Object}   response
   * A response for the transaction
   * @return {Promise<Object>}
   * A promise to be resolved with the signed token.
   */
  async function endTransaction(
    {
      id,
      req,
      res,
      delayPromise,
    }: {
      id: string;
      req: IncomingMessage;
      res: ServerResponse;
      delayPromise: Promise<void>;
    },
    response: WhookResponse,
    operationId = 'none',
  ): Promise<void> {
    /* Architecture Note #2.10.4: Transaction end
  We end the transaction by writing the final status
   and headers and piping the response body if any.

  The transaction can till error at that time but it
   is too late for changing the response status so
   we are just logging the event.
   This could be handled with
   [HTTP trailers](https://nodejs.org/api/http.html#http_response_addtrailers_headers)
   but the lack of client side support for now is
   preventing us to use them.

   Once terminated, the transaction is removed
    from the `TRANSACTIONS` hash.
  */
    try {
      await new Promise((resolve, reject) => {
        res.on('error', reject);
        res.on('finish', resolve);
        res.writeHead(
          response.status,
          statuses.message[response.status],
          Object.assign({}, response.headers || {}, { 'Transaction-Id': id }),
        );
        if (response.body && (response.body as Readable).pipe) {
          (response.body as Readable).pipe(res);
        } else {
          res.end();
        }
      });
    } catch (err) {
      FINAL_TRANSACTIONS[id].errored = true;
      apm('ERROR', {
        guruMeditation: id,
        request:
          FINAL_TRANSACTIONS[id].protocol +
          '://' +
          (req.headers.host || 'localhost') +
          FINAL_TRANSACTIONS[id].url,
        method: req.method as string,
        stack: printStackTrace(err as Error),
        operationId,
      });
    }
    FINAL_TRANSACTIONS[id].endTime = time();
    FINAL_TRANSACTIONS[id].endInBytes = req.socket.bytesRead;
    FINAL_TRANSACTIONS[id].endOutBytes = req.socket.bytesWritten;
    FINAL_TRANSACTIONS[id].statusCode = response.status;
    FINAL_TRANSACTIONS[id].resHeaders = response.headers || {};
    FINAL_TRANSACTIONS[id].operationId = operationId;

    apm('CALL', FINAL_TRANSACTIONS[id]);

    delete FINAL_TRANSACTIONS[id];

    try {
      await delay.clear(delayPromise);
    } catch (err) {
      log('debug', '❕ - Could not clear a delay.');
      log('debug-stack', printStackTrace(err as Error));
    }
  }
}
