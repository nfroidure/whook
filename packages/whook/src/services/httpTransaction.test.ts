/* eslint-disable @typescript-eslint/no-explicit-any */
import { describe, test, beforeEach, jest, expect } from '@jest/globals';
import StreamTest from 'streamtest';
import { YError } from 'yerror';
import initHTTPTransaction from './httpTransaction.js';
import { type WhookHandler, type WhookResponse } from './httpTransaction.js';
import { type WhookAPMService } from './apm.js';
import { type WhookObfuscatorService } from './obfuscator.js';
import { type IncomingMessage } from 'node:http';
import {
  type LogService,
  type TimeService,
  type DelayService,
} from 'common-services';

function streamifyBody(response) {
  return Object.assign({}, response, {
    body: StreamTest.v2.fromChunks([JSON.stringify(response.body)]),
  });
}

describe('initHTTPTransaction', () => {
  const log = jest.fn<LogService>();
  const apm = jest.fn<WhookAPMService>();
  const time = jest.fn<TimeService>();
  const uniqueId = jest.fn<() => string>();
  const delay = {
    create: jest.fn<DelayService['create']>(),
    clear: jest.fn<DelayService['clear']>(),
  };
  const obfuscator = {
    obfuscate: jest.fn<WhookObfuscatorService['obfuscate']>(),
    obfuscateSensibleProps:
      jest.fn<WhookObfuscatorService['obfuscateSensibleProps']>(),
    obfuscateSensibleHeaders:
      jest.fn<WhookObfuscatorService['obfuscateSensibleHeaders']>(),
  };

  beforeEach(() => {
    log.mockReset();
    apm.mockReset();
    time.mockReset();
    uniqueId.mockReset();
    delay.create.mockReset();
    delay.clear.mockReset();
    obfuscator.obfuscate.mockReset();
    obfuscator.obfuscateSensibleProps.mockReset();
    obfuscator.obfuscateSensibleHeaders.mockReset();
  });

  test('should work', async () => {
    const httpTransaction = await initHTTPTransaction({
      obfuscator,
      log,
      apm,
      time,
      delay,
      uniqueId,
    });

    expect(httpTransaction).toBeTruthy();
    expect({
      logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      apmCalls: apm.mock.calls,
      timeCalls: time.mock.calls,
      uniqueIdCalls: uniqueId.mock.calls,
      delayCreateCalls: delay.create.mock.calls,
      delayClearCalls: delay.clear.mock.calls,
      obfuscateCalls: obfuscator.obfuscate.mock.calls,
      obfuscateSensiblePropsCalls: obfuscator.obfuscateSensibleProps.mock.calls,
      obfuscateSensibleHeadersCalls:
        obfuscator.obfuscateSensibleHeaders.mock.calls,
    }).toMatchSnapshot();
  });

  describe('httpTransaction', () => {
    const buildResponse = jest.fn<any>();
    let res;
    let resBodyPromise;

    beforeEach((done) => {
      buildResponse.mockReset();
      resBodyPromise = new Promise((resolve, reject) => {
        res = StreamTest.v2.toText((err, text) => {
          if (err) {
            reject(err);
            return;
          }
          resolve(text);
        });
        res.writeHead = jest.fn();
        done();
      });
    });

    test('should work', async () => {
      time.mockReturnValue(new Date('2012-01-15T00:00:00.000Z').getTime());
      uniqueId.mockReturnValue('[id]');
      delay.create.mockReturnValueOnce(new Promise(() => undefined));
      delay.clear.mockResolvedValue(undefined);

      const httpTransaction = await initHTTPTransaction({
        obfuscator,
        log,
        apm,
        time,
        delay,
        uniqueId,
      });
      const req = {
        connection: { encrypted: true },
        ts: 1000000,
        ip: '127.0.0.1',
        method: 'GET',
        url: '/v1/users/1?extended=true',
        headers: {
          'x-forwarded-for': '127.0.0.1',
        },
        socket: {
          bytesRead: 16,
          bytesWritten: 64,
        },
      };

      buildResponse.mockResolvedValueOnce({
        body: StreamTest.v2.fromChunks([
          JSON.stringify({
            id: 1,
            name: 'John Doe',
          }),
        ]),
        headers: {
          'Content-Type': 'application/json',
        },
        status: 200,
      });

      const { request, transaction } = await httpTransaction(
        req as unknown as IncomingMessage,
        res,
      );

      const response = await transaction.start(
        buildResponse as unknown as WhookHandler,
      );

      await transaction.end(response, 'theOperationId');

      expect(request.url).toEqual('/v1/users/1?extended=true');
      expect(request.method).toEqual('get');
      expect(request.headers).toEqual({
        'x-forwarded-for': '127.0.0.1',
      });
      expect(request.body).toEqual(req);
      expect(res.writeHead.mock.calls).toEqual([
        [
          200,
          'OK',
          {
            'Content-Type': 'application/json',
            'Transaction-Id': '[id]',
          },
        ],
      ]);

      const text = await resBodyPromise;

      expect(text).toEqual('{"id":1,"name":"John Doe"}');
      expect({
        request,
        buildResponseCalls: buildResponse.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        apmCalls: apm.mock.calls,
        timeCalls: time.mock.calls,
        uniqueIdCalls: uniqueId.mock.calls,
        delayCreateCalls: delay.create.mock.calls,
        delayClearCalls: delay.clear.mock.calls,
        obfuscateCalls: obfuscator.obfuscate.mock.calls,
        obfuscateSensiblePropsCalls:
          obfuscator.obfuscateSensibleProps.mock.calls,
        obfuscateSensibleHeadersCalls:
          obfuscator.obfuscateSensibleHeaders.mock.calls,
      }).toMatchSnapshot();
    });

    test('should fail on timeout', async () => {
      time.mockReturnValue(new Date('2012-01-15T00:00:00.000Z').getTime());
      uniqueId.mockReturnValue('[id]');
      delay.create.mockResolvedValueOnce(undefined);
      delay.clear.mockResolvedValueOnce(undefined);

      let request;

      try {
        const httpTransaction = await initHTTPTransaction({
          obfuscator,
          log,
          apm,
          time,
          delay,
          uniqueId,
        });
        const req = {
          connection: { encrypted: true },
          ts: 1000000,
          ip: '127.0.0.1',
          method: 'GET',
          url: '/v1/users/1?extended=true',
          headers: {
            'x-forwarded-for': '127.0.0.1',
          },
          socket: {
            bytesRead: 16,
            bytesWritten: 64,
          },
        };

        buildResponse.mockReturnValueOnce(new Promise(() => undefined));

        const { request: _request, transaction } = await httpTransaction(
          req as unknown as IncomingMessage,
          res,
        );

        request = _request;

        await transaction.start(buildResponse as unknown as WhookHandler);

        await transaction.end({} as WhookResponse);

        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect((err as any).code).toEqual('E_TRANSACTION_TIMEOUT');
        expect((err as any).httpCode).toEqual(504);
        expect({
          request,
          logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
          apmCalls: apm.mock.calls,
          timeCalls: time.mock.calls,
          uniqueIdCalls: uniqueId.mock.calls,
          delayCreateCalls: delay.create.mock.calls,
          delayClearCalls: delay.clear.mock.calls,
          obfuscateCalls: obfuscator.obfuscate.mock.calls,
          obfuscateSensiblePropsCalls:
            obfuscator.obfuscateSensibleProps.mock.calls,
          obfuscateSensibleHeadersCalls:
            obfuscator.obfuscateSensibleHeaders.mock.calls,
        }).toMatchSnapshot();
      }
    });

    test('should fail with non-unique transaction id', async () => {
      time.mockReturnValue(new Date('2012-01-15T00:00:00.000Z').getTime());
      uniqueId.mockReturnValue('[id]');
      delay.create.mockReturnValueOnce(new Promise(() => undefined));
      delay.clear.mockResolvedValue(undefined);

      try {
        const httpTransaction = await initHTTPTransaction({
          obfuscator,
          log,
          apm,
          time,
          delay,
          uniqueId,
          TRANSACTIONS: { lol: {} },
        });
        const req = {
          connection: { encrypted: true },
          ts: 1000000,
          ip: '127.0.0.1',
          method: 'GET',
          url: '/v1/users/1?extended=true',
          headers: {
            'x-forwarded-for': '127.0.0.1',
            'transaction-id': 'lol',
          },
          socket: {
            bytesRead: 16,
            bytesWritten: 64,
          },
        };

        buildResponse.mockReturnValueOnce(Promise.resolve());

        const { transaction } = await httpTransaction(
          req as unknown as IncomingMessage,
          res,
        );

        await transaction
          .start(buildResponse as unknown as WhookHandler)
          .catch(transaction.catch)
          .then(streamifyBody)
          .then(transaction.end);

        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect((err as any).code).toEqual('E_TRANSACTION_ID_NOT_UNIQUE');
        expect((err as any).httpCode).toEqual(400);
        expect({
          logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
          timeCalls: time.mock.calls,
          uniqueIdCalls: uniqueId.mock.calls,
          delayCreateCalls: delay.create.mock.calls,
          delayClearCalls: delay.clear.mock.calls,
          obfuscateCalls: obfuscator.obfuscate.mock.calls,
          obfuscateSensiblePropsCalls:
            obfuscator.obfuscateSensibleProps.mock.calls,
          obfuscateSensibleHeadersCalls:
            obfuscator.obfuscateSensibleHeaders.mock.calls,
        }).toMatchSnapshot();
      }
    });
  });
});
