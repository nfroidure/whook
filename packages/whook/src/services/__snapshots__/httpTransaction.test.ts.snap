// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initHTTPTransaction httpTransaction should fail on timeout 1`] = `
{
  "apmCalls": [],
  "delayClearCalls": [],
  "delayCreateCalls": [
    [
      30000,
    ],
  ],
  "logCalls": [
    [
      "debug",
      "ðŸ’± - HTTP Transaction initialized.",
    ],
  ],
  "obfuscateCalls": [],
  "obfuscateSensibleHeadersCalls": [
    [
      {
        "x-forwarded-for": "127.0.0.1",
      },
    ],
  ],
  "obfuscateSensiblePropsCalls": [],
  "request": {
    "body": {
      "connection": {
        "encrypted": true,
      },
      "headers": {
        "x-forwarded-for": "127.0.0.1",
      },
      "ip": "127.0.0.1",
      "method": "GET",
      "socket": {
        "bytesRead": 16,
        "bytesWritten": 64,
      },
      "ts": 1000000,
      "url": "/v1/users/1?extended=true",
    },
    "headers": {
      "x-forwarded-for": "127.0.0.1",
    },
    "method": "get",
    "url": "/v1/users/1?extended=true",
  },
  "timeCalls": [
    [],
  ],
  "uniqueIdCalls": [
    [],
  ],
}
`;

exports[`initHTTPTransaction httpTransaction should fail with non-unique transaction id 1`] = `
{
  "delayClearCalls": [],
  "delayCreateCalls": [
    [
      30000,
    ],
  ],
  "logCalls": [
    [
      "debug",
      "ðŸ’± - HTTP Transaction initialized.",
    ],
  ],
  "obfuscateCalls": [],
  "obfuscateSensibleHeadersCalls": [
    [
      {
        "transaction-id": "lol",
        "x-forwarded-for": "127.0.0.1",
      },
    ],
  ],
  "obfuscateSensiblePropsCalls": [],
  "timeCalls": [
    [],
  ],
  "uniqueIdCalls": [
    [],
  ],
}
`;

exports[`initHTTPTransaction httpTransaction should work 1`] = `
{
  "apmCalls": [
    [
      "CALL",
      {
        "endInBytes": 16,
        "endOutBytes": 64,
        "endTime": 1326585600000,
        "errored": false,
        "id": "[id]",
        "ip": "127.0.0.1",
        "method": "GET",
        "operationId": "theOperationId",
        "protocol": "http",
        "reqHeaders": undefined,
        "resHeaders": {
          "Content-Type": "application/json",
        },
        "startInBytes": 16,
        "startOutBytes": 64,
        "startTime": 1326585600000,
        "statusCode": 200,
        "url": "/v1/users/1?extended=true",
      },
    ],
  ],
  "buildResponseCalls": [
    [],
  ],
  "delayClearCalls": [
    [
      Promise {},
    ],
  ],
  "delayCreateCalls": [
    [
      30000,
    ],
  ],
  "logCalls": [
    [
      "debug",
      "ðŸ’± - HTTP Transaction initialized.",
    ],
  ],
  "obfuscateCalls": [],
  "obfuscateSensibleHeadersCalls": [
    [
      {
        "x-forwarded-for": "127.0.0.1",
      },
    ],
  ],
  "obfuscateSensiblePropsCalls": [],
  "request": {
    "body": {
      "connection": {
        "encrypted": true,
      },
      "headers": {
        "x-forwarded-for": "127.0.0.1",
      },
      "ip": "127.0.0.1",
      "method": "GET",
      "socket": {
        "bytesRead": 16,
        "bytesWritten": 64,
      },
      "ts": 1000000,
      "url": "/v1/users/1?extended=true",
    },
    "headers": {
      "x-forwarded-for": "127.0.0.1",
    },
    "method": "get",
    "url": "/v1/users/1?extended=true",
  },
  "timeCalls": [
    [],
    [],
  ],
  "uniqueIdCalls": [
    [],
  ],
}
`;

exports[`initHTTPTransaction should work 1`] = `
{
  "apmCalls": [],
  "delayClearCalls": [],
  "delayCreateCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ’± - HTTP Transaction initialized.",
    ],
  ],
  "obfuscateCalls": [],
  "obfuscateSensibleHeadersCalls": [],
  "obfuscateSensiblePropsCalls": [],
  "timeCalls": [],
  "uniqueIdCalls": [],
}
`;
